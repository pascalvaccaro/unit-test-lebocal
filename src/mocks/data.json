{
  "lat": 43.7102,
  "lon": 7.262,
  "timezone": "Europe/Paris",
  "timezone_offset": 3600,
  "daily": [
      {
          "dt": 1639393200,
          "sunrise": 1639378582,
          "sunset": 1639410850,
          "moonrise": 1639400640,
          "moonset": 1639357740,
          "moon_phase": 0.33,
          "temp": {
              "day": 14.11,
              "min": 8.78,
              "max": 14.89,
              "night": 9.52,
              "eve": 11.46,
              "morn": 8.78
          },
          "feels_like": {
              "day": 12.8,
              "night": 8.54,
              "eve": 10.02,
              "morn": 7.47
          },
          "pressure": 1024,
          "humidity": 47,
          "dew_point": 2.98,
          "wind_speed": 2.69,
          "wind_deg": 307,
          "wind_gust": 3.41,
          "weather": [
              {
                  "id": 800,
                  "main": "Clear",
                  "description": "clear sky",
                  "icon": "01d"
              }
          ],
          "clouds": 8,
          "pop": 0,
          "uvi": 1.22
      },
      {
          "dt": 1639479600,
          "sunrise": 1639465028,
          "sunset": 1639497260,
          "moonrise": 1639488240,
          "moonset": 1639447920,
          "moon_phase": 0.36,
          "temp": {
              "day": 13.33,
              "min": 9.2,
              "max": 13.54,
              "night": 10.75,
              "eve": 11.17,
              "morn": 9.23
          },
          "feels_like": {
              "day": 11.79,
              "night": 9.34,
              "eve": 9.73,
              "morn": 7.99
          },
          "pressure": 1025,
          "humidity": 41,
          "dew_point": -0.59,
          "wind_speed": 2.57,
          "wind_deg": 335,
          "wind_gust": 2.49,
          "weather": [
              {
                  "id": 800,
                  "main": "Clear",
                  "description": "light rain",
                  "icon": "01d"
              }
          ],
          "clouds": 0,
          "pop": 0,
          "uvi": 1.1
      },
      {
          "dt": 1639566000,
          "sunrise": 1639551472,
          "sunset": 1639583672,
          "moonrise": 1639575960,
          "moonset": 1639538100,
          "moon_phase": 0.39,
          "temp": {
              "day": 14.5,
              "min": 10.03,
              "max": 14.56,
              "night": 10.98,
              "eve": 11.93,
              "morn": 10.11
          },
          "feels_like": {
              "day": 13,
              "night": 9.54,
              "eve": 10.48,
              "morn": 8.48
          },
          "pressure": 1027,
          "humidity": 38,
          "dew_point": -0.44,
          "wind_speed": 2.67,
          "wind_deg": 339,
          "wind_gust": 2.85,
          "weather": [
              {
                  "id": 800,
                  "main": "Clear",
                  "description": "clear sky",
                  "icon": "01d"
              }
          ],
          "clouds": 9,
          "pop": 0,
          "uvi": 1.19
      },
      {
          "dt": 1639652400,
          "sunrise": 1639637914,
          "sunset": 1639670086,
          "moonrise": 1639663920,
          "moonset": 1639628340,
          "moon_phase": 0.42,
          "temp": {
              "day": 13.68,
              "min": 10.05,
              "max": 13.68,
              "night": 10.67,
              "eve": 11.01,
              "morn": 10.14
          },
          "feels_like": {
              "day": 12.17,
              "night": 9.31,
              "eve": 9.63,
              "morn": 8.46
          },
          "pressure": 1028,
          "humidity": 41,
          "dew_point": -0.51,
          "wind_speed": 2.61,
          "wind_deg": 340,
          "wind_gust": 2.64,
          "weather": [
              {
                  "id": 801,
                  "main": "Clouds",
                  "description": "few clouds",
                  "icon": "02d"
              }
          ],
          "clouds": 14,
          "pop": 0,
          "uvi": 1.18
      },
      {
          "dt": 1639738800,
          "sunrise": 1639724354,
          "sunset": 1639756503,
          "moonrise": 1639752060,
          "moonset": 1639718520,
          "moon_phase": 0.45,
          "temp": {
              "day": 13.79,
              "min": 9.76,
              "max": 13.79,
              "night": 10.5,
              "eve": 10.96,
              "morn": 9.76
          },
          "feels_like": {
              "day": 12.3,
              "night": 9.07,
              "eve": 9.68,
              "morn": 8.51
          },
          "pressure": 1027,
          "humidity": 41,
          "dew_point": -0.15,
          "wind_speed": 2.54,
          "wind_deg": 326,
          "wind_gust": 2.71,
          "weather": [
              {
                  "id": 800,
                  "main": "Clear",
                  "description": "clear sky",
                  "icon": "01d"
              }
          ],
          "clouds": 1,
          "pop": 0,
          "uvi": 1.16
      },
      {
          "dt": 1639825200,
          "sunrise": 1639810791,
          "sunset": 1639842922,
          "moonrise": 1639840620,
          "moonset": 1639808640,
          "moon_phase": 0.48,
          "temp": {
              "day": 12.18,
              "min": 8.63,
              "max": 12.18,
              "night": 8.63,
              "eve": 9.67,
              "morn": 8.93
          },
          "feels_like": {
              "day": 11.02,
              "night": 6.74,
              "eve": 6.49,
              "morn": 7.29
          },
          "pressure": 1020,
          "humidity": 60,
          "dew_point": 3.86,
          "wind_speed": 8.7,
          "wind_deg": 100,
          "wind_gust": 9.68,
          "weather": [
              {
                  "id": 800,
                  "main": "Clear",
                  "description": "clear sky",
                  "icon": "01d"
              }
          ],
          "clouds": 2,
          "pop": 0.19,
          "uvi": 2
      },
      {
          "dt": 1639911600,
          "sunrise": 1639897227,
          "sunset": 1639929344,
          "moonrise": 1639929600,
          "moonset": 1639898520,
          "moon_phase": 0.5,
          "temp": {
              "day": 8.93,
              "min": 5.25,
              "max": 8.94,
              "night": 5.25,
              "eve": 6.07,
              "morn": 6.54
          },
          "feels_like": {
              "day": 5.18,
              "night": 1.94,
              "eve": 2.64,
              "morn": 2.8
          },
          "pressure": 1027,
          "humidity": 49,
          "dew_point": -1.94,
          "wind_speed": 8.37,
          "wind_deg": 98,
          "wind_gust": 13.96,
          "weather": [
              {
                  "id": 500,
                  "main": "Rain",
                  "description": "light rain",
                  "icon": "10d"
              }
          ],
          "clouds": 35,
          "pop": 0.72,
          "rain": 3.2,
          "uvi": 2
      },
      {
          "dt": 1639998000,
          "sunrise": 1639983660,
          "sunset": 1640015768,
          "moonrise": 1640019060,
          "moonset": 1639988160,
          "moon_phase": 0.54,
          "temp": {
              "day": 10.62,
              "min": 4.77,
              "max": 10.62,
              "night": 6.15,
              "eve": 6.89,
              "morn": 4.99
          },
          "feels_like": {
              "day": 8.21,
              "night": 3.1,
              "eve": 4.65,
              "morn": 1.15
          },
          "pressure": 1024,
          "humidity": 18,
          "dew_point": -14.01,
          "wind_speed": 5.37,
          "wind_deg": 4,
          "wind_gust": 10.91,
          "weather": [
              {
                  "id": 803,
                  "main": "Clouds",
                  "description": "broken clouds",
                  "icon": "04d"
              }
          ],
          "clouds": 83,
          "pop": 0,
          "uvi": 2
      }
  ]
}