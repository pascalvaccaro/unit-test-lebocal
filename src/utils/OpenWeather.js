export const openWeather = (res) => (res && res.daily[1] ? res.daily[1].weather : [])
  .reduce((acc, weather) => {
    if (weather.description === "clear sky") acc.sunny = true;
    if (weather.description.includes("rain")) acc.raining = true;
    return acc;
  }, {});