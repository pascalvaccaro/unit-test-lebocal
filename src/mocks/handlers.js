import { rest } from 'msw';
import * as data from './data.json';

export const handlers = [
  rest.get("https://api.openweathermap.org/data/2.5/onecall", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json(data),
      ctx.delay(),
      // ctx.delay('infinite'),
      // ctx.delay()
    )
  })
];