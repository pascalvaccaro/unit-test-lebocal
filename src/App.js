import { useEffect, useState } from 'react';
import axios from 'axios';
import './App.css';
import Weather from './components/Weather';
import { openWeather } from './utils/OpenWeather';
// 0. Axios must return something in the response
// 0. openWeather must return an object
// 1. Test with nothing
// 2. Test with data

function App({ lat, lon}) {
  const [data, setData] = useState({});
  const [error, setError] = useState();

  useEffect(() => {
    if(lat >= 0 && lon >= 0) {
      const url = new URL(process.env.REACT_APP_OWM_ONE_CALL, process.env.REACT_APP_OWM_API_HOST);
      url.searchParams.set('appid', process.env.REACT_APP_OWM_API_KEY);
      url.searchParams.set('lat', lat);
      url.searchParams.set('lon', lon);
      axios.get(url.toString())
        .then(res => res.data)
        .then(openWeather)
        .then(response => setData(response))
        .catch(err => setError(err));
    }
  }, [lat, lon]);

  return (
    <div className="App">
      <header className="App-header">
        {!error ? <Weather {...data} /> : null}
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
        <a
          className="App-link"
          href="https://openweathermap.org/"
          target="_blank"
          rel="noopener noreferrer"
        >
          Is it going to rain tomorrow?
        </a>
        <button onClick={() => setData({ sunny: true })}>Sunny</button>
      </header>
    </div>
  );
}

export default App;
