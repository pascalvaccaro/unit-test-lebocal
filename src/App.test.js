import { act, waitFor, render, screen, fireEvent } from '@testing-library/react';
import App from './App';
import { rest } from 'msw';
import { worker } from './mocks/browser';
import axios from 'axios';
// Create a spy
import { openWeather } from './utils/OpenWeather';

jest.mock("axios");
// -> (load this mock instead of the real module)
jest.mock("./utils/OpenWeather");

describe('App', () => {
  let result;
  beforeEach(() => {
    act(() => { result = render(<App />); });
    // Fill the Axios spy with a mocked (dummy) implementation
    axios.get.mockResolvedValue(null);
    openWeather.mockReturnValue({ sunny: true });
  });
  
  afterEach(() => {
    openWeather.mockReset();
  });
  
  it('renders Is it going to rain tomorrow?', () => {
    const linkElement = screen.getByText(/Is it going to rain tomorrow?/i);
    return waitFor(() => expect(linkElement).toBeInTheDocument());
  });
  
  test('renders Open Weather Link', (done) => {
    const linkElement = screen.getByText(/Is it going to rain tomorrow?/i);
    waitFor(() => expect(linkElement.href).toBe('https://openweathermap.org/'))
      .then(done);
  });
  
  test('axios is invoked with the right parameters', async () => {
    await waitFor(() => {
      expect(axios.get).not.toHaveBeenCalled();
    });

    result.rerender(<App lat={0} lon={0} />);

    await waitFor(() => {
      expect(axios.get).toHaveBeenCalledTimes(1)
      expect(axios.get).toHaveBeenCalledWith(
        "https://api.openweathermap.org/data/2.5/onecall?appid=ce090da8d558e94a550b538120d1f823&lat=0&lon=0"
      );
    });
  });

  describe.skip('Real API calls with Axios', () => {
    beforeEach(() => {
      axios.mockRestore();
      openWeather.mockRestore();
    });
    
    test('the openweather API is requested and it is raining', async () => {
      result.rerender(<App lat={0} lon={0} />);
      await waitFor(() => expect(screen.getByAltText(/logo/).src).toMatch(/rain\.svg$/));
    });
  
    test('the weather is not displayed when an error occurs', async () => {
      expect(screen.getByAltText(/logo/)).toBeInTheDocument();
      worker.use(
        rest.get("https://api.openweathermap.org/data/2.5/onecall", (req, res, ctx) => {
          return res(
            ctx.status(404),
            ctx.json({ error: "No town found here"})
          );
        })
      );
      result.rerender(<App lat={0} lon={0} />);
      await waitFor(() => expect(screen.queryByAltText(/logo/)).toBe(null));
    });
  })

  describe('The cheat sunny button', () => {
    test('a button is present with the "Sunny" label', () => {
      const btnElement = screen.getByText(/Sunny/);
      expect(btnElement).toBeInTheDocument();
    });
    
    test('clicking the sunny button should set the Weather to sunny', async () => {
      const btnElement = result.getByText(/Sunny/);
      fireEvent.click(btnElement);
      const imgElement = screen.getByAltText(/logo/);
      expect(imgElement).toMatchSnapshot();
    });
  });

});
