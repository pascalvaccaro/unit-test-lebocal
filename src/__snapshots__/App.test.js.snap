// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App The cheat sunny button clicking the sunny button should set the Weather to sunny 1`] = `
<img
  alt="logo"
  src="sun.svg"
/>
`;
